lib/config/configuration_test.go

```diff
@@ -2156,3 +2172,135 @@ func TestTLSCert(t *testing.T) {
 		})
 	}
 }
+
+func TestApplyKeyStoreConfig(t *testing.T) {
+	slotNumber := 1
+
+	tempDir := t.TempDir()
+
+	worldReadablePinFilePath := filepath.Join(tempDir, "world-readable-pin-file")
+	err := os.WriteFile(worldReadablePinFilePath, []byte("world-readable-pin-file"), 0o644)
+	require.NoError(t, err)
+	securePinFilePath := filepath.Join(tempDir, "secure-pin-file")
+	err = os.WriteFile(securePinFilePath, []byte("secure-pin-file"), 0o600)
+	require.NoError(t, err)
+
+	worldWritablePKCS11LibPath := filepath.Join(tempDir, "world-writable-pkcs1")
+	err = os.WriteFile(worldWritablePKCS11LibPath, []byte("pkcs11"), 0o666)
+	require.NoError(t, err)
+	require.NoError(t, os.Chmod(worldWritablePKCS11LibPath, 0o666))
+	securePKCS11LibPath := filepath.Join(tempDir, "secure-pkcs11")
+	err = os.WriteFile(securePKCS11LibPath, []byte("pkcs11"), 0o600)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name string
+
+		auth Auth
+
+		want    keystore.Config
+		wantErr string
+	}{
+		{
+			name: "handle nil configuration",
+			auth: Auth{
+				CAKeyParams: nil,
+			},
+			want: service.MakeDefaultConfig().Auth.KeyStore,
+		},
+		{
+			name: "correct config",
+			auth: Auth{
+				CAKeyParams: &CAKeyParams{
+					PKCS11: PKCS11{
+						ModulePath: securePKCS11LibPath,
+						TokenLabel: "foo",
+						SlotNumber: &slotNumber,
+						Pin:        "pin",
+					},
+				},
+			},
+			want: keystore.Config{
+				TokenLabel: "foo",
+				SlotNumber: &slotNumber,
+				Pin:        "pin",
+				Path:       securePKCS11LibPath,
+			},
+		},
+		{
+			name: "correct config with pin file",
+			auth: Auth{
+				CAKeyParams: &CAKeyParams{
+					PKCS11: PKCS11{
+						ModulePath: securePKCS11LibPath,
+						TokenLabel: "foo",
+						SlotNumber: &slotNumber,
+						PinPath:    securePinFilePath,
+					},
+				},
+			},
+			want: keystore.Config{
+				TokenLabel: "foo",
+				SlotNumber: &slotNumber,
+				Pin:        "secure-pin-file",
+				Path:       securePKCS11LibPath,
+			},
+		},
+		{
+			name: "err when pin and pin path configured",
+			auth: Auth{
+				CAKeyParams: &CAKeyParams{
+					PKCS11: PKCS11{
+						Pin:     "oops",
+						PinPath: securePinFilePath,
+					},
+				},
+			},
+			wantErr: "can not set both pin and pin_path",
+		},
+		{
+			name: "err when pkcs11 world writable",
+			auth: Auth{
+				CAKeyParams: &CAKeyParams{
+					PKCS11: PKCS11{
+						ModulePath: worldWritablePKCS11LibPath,
+					},
+				},
+			},
+			wantErr: fmt.Sprintf(
+				"PKCS11 library (%s) must not be world-writable",
+				worldWritablePKCS11LibPath,
+			),
+		},
+		{
+			name: "err when pin file world-readable",
+			auth: Auth{
+				CAKeyParams: &CAKeyParams{
+					PKCS11: PKCS11{
+						PinPath: worldReadablePinFilePath,
+					},
+				},
+			},
+			wantErr: fmt.Sprintf(
```

Comment:
nit:
I'd rename it to `errMessage` or something similar as it's not a flag `want/don't want` but an error message that we want to assert.

Reply:
👍